language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/FromJsonGenerator.scala

```diff
@@ -0,0 +1,256 @@
+      .addModifiers(modifiers: _*)
+      .addTypeVariables(typeParams.map(TypeVariableName.get).asJava)
+      .addParameters(jsonDecoderParamsForTypeParams(typeParams))
+      .returns(decoderTypeName(typeName))
+      .addStatement(
+        "return $T.record($L, $L, $L)",
+        decodeClass,
+        fieldNames,
+        fieldsByName.toString(),
+        constr,
+      )
+      .build()
+  }
+
+  private def forRecordLikeFromString(
+      className: ClassName,
+      typeParams: IndexedSeq[String],
+  ): MethodSpec =
+    MethodSpec
+      .methodBuilder("fromJson")
+      .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+      .addTypeVariables(typeParams.map(TypeVariableName.get).asJava)
+      .addParameter(classOf[String], "json")
+      .addParameters(jsonDecoderParamsForTypeParams(typeParams))
+      .returns(className.parameterized(typeParams))
+      .addException(classOf[java.io.IOException])
+      .addStatement(
+        "return jsonDecoder($L).decode(new $T(json))",
+        CodeBlock.join(typeParams.map(t => CodeBlock.of(decodeTypeParamName(t))).asJava, ", "),
+        classOf[JsonLfReader],
+      )
+      .build()
+
+  def forVariant(
+      className: ClassName,
+      typeParams: IndexedSeq[String],
+      fields: Fields,
+  ) = {
+    val typeName = className.parameterized(typeParams)
+
+    val tagNames = CodeBlock.of(
+      "$T.asList($L)",
+      classOf[java.util.Arrays],
+      CodeBlock.join(fields.map(f => CodeBlock.of("$S", f.javaName)).asJava, ", "),
+    )
+    val variantsByTag = {
+      val block = CodeBlock
+        .builder()
+        .beginControlFlow("name ->")
+        .beginControlFlow("switch (name)")
+      fields.foreach { f =>
+        block.addStatement(
+          "case $S: return $L($L)",
+          f.damlName,
+          decoderForTagName(f.damlName),
+          CodeBlock.join(typeParams.map(t => CodeBlock.of(decodeTypeParamName(t))).asJava, ", "),
+        )
+      }
+      block
+        .addStatement("default: return null")
+        .endControlFlow() // end switch
+        .endControlFlow() // end lambda
+        .build()
+    }
+
+    MethodSpec
+      .methodBuilder("jsonDecoder")
+      .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+      .addTypeVariables(typeParams.map(TypeVariableName.get).asJava)
+      .addParameters(jsonDecoderParamsForTypeParams(typeParams))
+      .returns(decoderTypeName(typeName))
+      .addStatement("return $T.variant($L, $L)", decodeClass, tagNames, variantsByTag.toString())
+      .build()
+  }
+
+  def forVariantRecord(
+      tag: String,
+      fields: Fields,
+      className: ClassName,
+      typeParams: IndexedSeq[String],
+  )(implicit
+      packagePrefixes: PackagePrefixes
+  ) =
+    forRecordLike(
+      decoderForTagName(tag),
+      Seq(Modifier.PRIVATE, Modifier.STATIC),
+      fields,
+      className,
+      typeParams,
+    )
+
+  def forVariantSimple(typeName: TypeName, typeParams: IndexedSeq[String], field: FieldInfo)(
```

Comment:
not clear from the naming what a simple variant vs a variant record is

Reply:
hm the term I usually use is sums of products or SOP for short. But I'm fine with just adding a comment.